File: src/lib/format.ts
import { isNotCommentOrBlank } from "./helpers.ts"
import { calcColumnWidths, stringifyLayer } from "./layer.ts"
import { Block, Layout, LayoutError } from "./types.ts"

export function rewriteLines(lines: string[], blocks: Block[], layout: Layout) {
  const columnWidths = calcColumnWidths(layout.layers)
  blocks.filter((b) => b.name.startsWith("layer:")).forEach((layerBlock) => {
    const { lines: layerLines, name, startLineNr } = layerBlock
    const layerName = name.replace("layer:", "").trim()
    const layer = layout.layers.find((l) => l.name === layerName)
    if (layer == null) {
      throw new LayoutError(`This is weird, can't find a layer named '${layerName}`)
    }
    const formattedLines = stringifyLayer(layer, layout.structure, columnWidths)

    let j = 0
    for (let i = startLineNr - 1; i < startLineNr + layerLines.length - 1; i++) {
      if (isNotCommentOrBlank(lines[i])) {
        lines[i] = formattedLines[j++]
      }
    }
  })
}

File: internal/format/format.go
package format

import (
	"fmt"
	"strings"

	"github.com/elentok/qmkmd/internal/layout"
)

func RewriteLines(lines []string, blocks []layout.Block, parsed layout.Layout) error {
	columnWidths := layout.CalcColumnWidths(parsed.Layers)

	for _, layerBlock := range blocks {
		if !strings.HasPrefix(layerBlock.Name, "layer:") {
			continue
		}

		layerName := strings.TrimSpace(strings.TrimPrefix(layerBlock.Name, "layer:"))
		var current *layout.Layer
		for i := range parsed.Layers {
			if parsed.Layers[i].Name == layerName {
				current = &parsed.Layers[i]
				break
			}
		}
		if current == nil {
			return &layout.LayoutError{Msg: fmt.Sprintf("This is weird, can't find a layer named '%s'", layerName)}
		}

		formattedLines := layout.StringifyLayer(*current, parsed.Structure, columnWidths)
		j := 0
		for i := layerBlock.StartLineNr - 1; i < layerBlock.StartLineNr+len(layerBlock.Lines)-1; i++ {
			if layout.IsNotCommentOrBlank(lines[i]) {
				lines[i] = formattedLines[j]
				j++
			}
		}
	}

	return nil
}
